packages/field-base/test/input-constraints-mixin.test.js

```diff
@@ -44,27 +44,65 @@ const runTests = (baseClass) => {
   );
 
   describe('validation', () => {
-    let element, input, validateSpy, changeSpy;
+    let element, input, validateSpy;
 
-    beforeEach(() => {
-      element = document.createElement(tag);
-      validateSpy = sinon.spy(element, 'validate');
-    });
+    describe('initial', () => {
+      beforeEach(() => {
+        element = document.createElement(tag);
+        validateSpy = sinon.spy(element, 'validate');
+      });
 
-    afterEach(() => {
-      element.remove();
-    });
+      afterEach(() => {
+        element.remove();
+      });
 
-    describe('without value', () => {
-      beforeEach(async () => {
+      it('should not validate without value', async () => {
         document.body.appendChild(element);
         await nextRender();
+        expect(validateSpy.called).to.be.false;
       });
 
-      it('should not validate initially', () => {
+      it('should not validate without value when the field has invalid', async () => {
+        element.invalid = true;
+        document.body.appendChild(element);
+        await nextRender();
         expect(validateSpy.called).to.be.false;
       });
 
+      describe('with value', () => {
+        beforeEach(() => {
+          element.value = 'Value';
+        });
+
+        it('should not validate without constraints', async () => {
+          document.body.appendChild(element);
+          await nextRender();
+          expect(validateSpy.called).to.be.false;
+        });
+
+        it('should not validate without constraints when the field has invalid', async () => {
+          element.invalid = true;
+          document.body.appendChild(element);
+          await nextRender();
+          expect(validateSpy.called).to.be.false;
+        });
+
+        it('should validate when the field has a constraint', async () => {
```

Comment:
When run under Lit, this test fails due to validation occuring twice during initialization. 

Currently, the `_constraintsChanged` observer is called for `minlength` and `stateTarget` individually. When run under Polymer, `stateTarget` is not yet defined at the moment the observer is called for `minlength`, so the validation is skipped. However, in the case of Lit, property changes are batched, so it turns out that the observer is called for each property when `stateTarget` is already defined.

Reply:
Seems that the second validation occurs because [this controller](https://github.com/vaadin/web-components/blob/master/packages/field-base/test/input-constraints-mixin.test.js#L36-L41) is added in `ready()` (which runs after constructor).

Could possibly be fixed by adding it in the constructor instead. But this causes a timing issue in slot-controller, which could be tackled with:
![Screenshot 2022-08-30 at 11 00 33](https://user-images.githubusercontent.com/1222264/187383121-3942e445-1842-40d4-8973-35de51f3167f.png)

Reply:
Registering controllers in the constructor is an option that I was thinking about too. However, what I can't understand is why Lit triggers observers twice in the above case at all. Shouldn't it batch them into a single call after all?

Reply:
PolylitMixin runs all observers in `updated()` https://github.com/vaadin/web-components/blob/master/packages/component-base/src/polylit-mixin.js#L187

...but it only calls `ready()` after the first update is finished https://github.com/vaadin/web-components/blob/master/packages/component-base/src/polylit-mixin.js#L183

This causes two validation rounds in the test case

Reply:
Ok, but then I guess `stateTarget` should be undefined during the first update and the first validation should not happen.

Reply:
> Ok, but then I guess stateTarget should be undefined during the first update (as with Polymer) so the first validation should not occur, but it occurs somehow

True, it actually seems to be defined during the first `update()` even though it's not listed in the changed properties (`props`)...weird

Reply:
It is also weird that no properties are currently considered changed at `firstUpdated` (the `props` argument is an empty map). I would have expected that at least properties with initial values were reported as changed.

Reply:
Tried blocking `update()` until firstUpdate is finished, seems to work but not sure how solid it is as a workaround:

![Screenshot 2022-08-30 at 12 11 09](https://user-images.githubusercontent.com/1222264/187398151-f31daece-9f46-4119-a7fa-c4522ce17a0e.png)

Reply:
Would that work when no properties are changed in the `ready` callback?

Reply:
Didn't try, if necessary, the microtask could also contain a (conditional) `requestUpdate()` to make sure `updated()` gets run once after the first update